includes/class-wc-gateway-komoju-ipn-handler.php

```diff
@@ -4,76 +4,93 @@
 }
 
 include_once( 'class-wc-gateway-komoju-response.php' );
+include_once( 'class-wc-gateway-komoju-webhook-event.php');
 
 /**
  * Handles responses from Komoju IPN
  */
 class WC_Gateway_Komoju_IPN_Handler extends WC_Gateway_Komoju_Response {
 
-	protected $notify_url;
+	protected $webhookSecretToken;
 	protected $secret_key;
 	protected $invoice_prefix;
 	/**
 	 * Constructor
 	 */
-	public function __construct( $sandbox = false, $notify_url = '', $secret_key = '', $invoice_prefix = ''  ) {
+	public function __construct( $webhookSecretToken = '', $secret_key = '', $invoice_prefix = ''  ) {
 		add_action( 'woocommerce_api_wc_gateway_komoju', array( $this, 'check_response' ) );
 		add_action( 'valid-komoju-standard-ipn-request', array( $this, 'valid_response' ) );
 
-		$this->sandbox        	= $sandbox;
-		$this->notify_url	  	= $notify_url;
-		$this->secret_key	  	= $secret_key;
-		$this->invoice_prefix	= $invoice_prefix;
+		$this->webhookSecretToken	  	= $webhookSecretToken;
+		$this->secret_key	   	        = $secret_key;
+		$this->invoice_prefix			= $invoice_prefix;
 	}
 
 	/**
 	 * Check for Komoju IPN Response
 	 */
 	public function check_response() {
-		if ( ! empty( $_GET ) && $this->validate_hmac() ) {
-			$posted = wp_unslash( $_GET );
+		$entityBody = file_get_contents('php://input');
 
-			do_action( "valid-komoju-standard-ipn-request", $posted['transaction'] );
+		if ( ! empty( $entityBody ) && $this->validate_hmac($entityBody) ) {
+			$webhookEvent = new WC_Gateway_Komoju_Webhook_Event($entityBody);
+
+			do_action( "valid-komoju-standard-ipn-request", $webhookEvent );
 			exit;
 		}
 		wp_die( "Komoju IPN Request Failure", "Komoju IPN", array( 'response' => 500 ) );
 	}
 
 	/**
 	 * There was a valid response
-	 * @param  array $posted Post data after wp_unslash
+	 * @param  WC_Gateway_Komoju_Webhook_Event $webhookEvent Webhook event data
 	 */
-	public function valid_response( $posted ) {
-		WC_Gateway_Komoju::log( 'External order num: ' . $posted['external_order_num'] );
-		WC_Gateway_Komoju::log( 'Uuid: ' . $posted['uuid'] );
-		WC_Gateway_Komoju::log( 'Payment status: ' . $posted['status'] );
+	public function valid_response( $webhookEvent ) {
+		WC_Gateway_Komoju::log( 'External order num: ' . $webhookEvent->external_order_num() );
+		WC_Gateway_Komoju::log( 'Uuid: ' . $webhookEvent->uuid() );
+		WC_Gateway_Komoju::log( 'Payment status: ' . $webhookEvent->status() );
 
-		$order = $this->get_komoju_order( $posted, $this->invoice_prefix );
+		$order = $this->get_komoju_order( $webhookEvent, $this->invoice_prefix );
 		if ( $order ) {
-			if ( method_exists( $this, 'payment_status_' . $posted['status'] ) ) {
-				call_user_func( array( $this, 'payment_status_' . $posted['status'] ), $order, $posted );
+			switch($webhookEvent->status()) {

```

Comment:
I replaced the meta programming with a case statement to make it clearer as to which method was being called. I didn't feel like the meta programming was benefitting us much, as we would have still had to come to this file and update it with a new method every time we wanted to support a new event type, so it wasn't really saving us much time, at the cost of being harder to read and harder to navigate the codebase (doing a global search for any of the `payment_status_*` methods would return nothing, making it hard to understand where the method was being called from).

I understand this is just personal preference on my part, so if people want to switch back to the `call_user_func` method, I'm happy to do that 😄